### ISSUE TITLE
binary detection: openbsd OpenSSH and portable OpenSSH

### ISSUE THREAD

**Issue by krysgor** [anchore/syft#3216]
Hi,

(Not sure if i'm right here, because it's a contributor question and i'm not so familiar with go)

I would like to implement openbsd OpenSSH and portable OpenSSH binary detection with correct cpe's in one classifier.

So openbsd have two OpenSSH products with different cpe's:
- [normal](https://nvd.nist.gov/products/cpe/detail/DE876FD6-4D26-463F-886E-6A53DE98835B?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aopenbsd%3Aopenssh&status=FINAL) `cpe:2.3:a:openbsd:openssh:9.6:-:*:*:*:*:*:*`
- [portable](https://nvd.nist.gov/products/cpe/detail/64848251-5FCD-476A-B2AB-744B23F6243F?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aopenbsd%3Aopenssh&status=FINAL) `cpe:2.3:a:openbsd:openssh:9.6:p1:*:*:*:*:*:*`

I alrady have the regex to match the version `\x00OpenSSH_(?P<version>[0-9]+\.[0-9]+)(p[0-9])?\x00` (is also match the optional portable `p1` information).

The question ist: how can I build this two different cpe in one classifier?
Is it possible to implement this with one classifier? If not I will make simply two classifyers: openssh-binary and openssh-portable-binary.

Thanks

**Comment by spiffcs** [comment]
👋 Thanks for the issue @krysgor and question about classifiers!. 

I'll point you to a recent PR that just landed that has a few examples in it:
https://github.com/anchore/syft/pull/3078

Note the classifiers being added here:
https://github.com/anchore/syft/pull/3078/files#diff-962c0bf8d15912f4f2b27bb43392f4ec0ab0d535cd5848ec6da96e8c251f9017R450-R479

On the `Classifier` struct there is a field called `CPEs`. While this PR uses a convenience function for `singleCPE` that field is flexible to contain multiple CPE so long as Class/Package match is the same.

If you wanted two classifiers that resulted in two different packages you would just use a single cpe and do something like:
```
		{
			Class:    "openbsd-OpenSSH-binary",
			Package: "openSSH/openbsd",
			CPEs:    singleCPE("cpe:2.3:a:openbsd:openssh:9.6:-:*:*:*:*:*:*"),
		},
		{
			Class:    "portable-OpenSSH",
			Package: "openSSH/portable",
			CPEs:    singleCPE("cpe:2.3:a:openbsd:openssh:9.6:p1:*:*:*:*:*:*"),
		},
```

The other path is to just do this as one classifier with one package where you add multiple CPE at the bottom of that struct.

Happy to talk more about this or review some code if you've already written something 👍

**Comment by wagoodman** [comment]
> But creating an sbom that contains two entries for openssh is (probably) wrong

agreed -- mind posting the code for the two regexes that were used in the dual-classifier approach? There might be more options, I think you really need one classifier with multiple evidence matchers, see an example here

https://github.com/anchore/syft/blob/d7005d7d8ca6d05f594f7bc1a140ae1e85bc0328/syft/pkg/cataloger/binary/classifiers.go#L13-L22

This way we would never be finding duplicate packages since there would be one classifier.
